src/migrate/index.js

```diff
@@ -121,22 +138,89 @@ export default class Migrator {
     });
   }
 
+  _createMigrationLockTable(tableName) {
+    return this.knex.schema.createTable(tableName, function(t) {
+      t.integer('is_locked');
+    });
+  }
+
+  _getLockTableName() {
+    return this.config.tableName + '_lock';
+  }
+
+  _isLocked(trx) {
+    var tableName = this._getLockTableName();
+    return this.knex(tableName)
+      .transacting(trx)
+      .forUpdate()
+      .select('*')
+      .then(data => data[0].is_locked);
+  }
+
+  _lockMigrations(trx) {
+    var tableName = this._getLockTableName();
+    return this.knex(tableName)
+      .transacting(trx)
+      .update({ is_locked: 1 });
+  }
+
+  _getLock() {
+    return this.knex.transaction(trx => {
+      return this._isLocked(trx)
+        .then(isLocked => {
+          if (isLocked) {
+            throw new Error("Migration table is are already locked");
+          }
+        })
+        .then(() => this._lockMigrations(trx));
+    });
+  }
+
+  _freeLock() {
+    var tableName = this._getLockTableName();
+    return this.knex(tableName)
+      .update({ is_locked: 0 });
+  }
+
   // Run a batch of current migrations, in sequence.
   _runBatch(migrations, direction) {
-    return Promise.all(_.map(migrations, this._validateMigrationStructure, this))
-      .then(() => this._latestBatchNumber())
-      .then((batchNo) => {
-        if (direction === 'up') batchNo++;
-        return batchNo;
-      })
-      .then((batchNo) => {
-        return this._waterfallBatch(batchNo, migrations, direction)
-      })
-      .catch((error) => {
+    return this._getLock()
+    .catch(err => {
```

Comment:
This `catch` statement could be moved into `_getLock` (after transaction handler) - feels more sensible to fire the `LockError` from there. (or perhaps I'm just bike shedding)